### PR TITLE
KAFKA-8220 & KIP-345 part-3: Avoid kicking out members through rebalance timeout

### PR DESCRIPTION
**Pull request by abbccdda**
To make consumer group members more persist, we want to avoid kick-out unjoined members through rebalance timeout. The only exception is when leader fails to join, because we will at risk of no assignment computed during sync stage. The choice will be kicking off non-responsive leader and choose a new leader if possible.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by abbccdda** [comment]
Retest this please

**Comment by abbccdda** [comment]
@guozhangwang @hachikuji when you got time :)

**Review by rajinisivaram — COMMENTED** [review]
@abbccdda Thanks for the PR. Changes look good to me, but it will be good to get a review from @hachikuji before merging in case I missed something.

**Review by rajinisivaram — COMMENTED** [review]
@abbccdda Thanks for the PR. Changes look good to me, but it will be good to get a review from @hachikuji before merging in case I missed something.

**Comment by abbccdda** [comment]
@rajinisivaram Thanks a lot for the review! Yea, pinging @hachikuji @guozhangwang again since we got your green light lol.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: Could use `group.get(leaderId)`?

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: If I understand correctly, we are trying to ensure that we still have a leader that can do the assignment. But is it necessary to actually remove the leader from the group? It makes the behavior a little inconsistent. Could we just change the leader?

**Review by hachikuji — COMMENTED** [review]
Thanks @abbccdda. I just had a question.

File: core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala]: You can use `foreach` instead of `match` if there is nothing to do in the `None` case

File: core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala]: Perhaps the name should convey the fact that we are choosing from the members which are awaiting join?

Comment (hachikuji) [reply core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala]: Perhaps `maybeElectJoinedLeader`? Another question I'm wondering is if it's possible to reach this state when _all_ the static members are offline?

**Comment by abbccdda** [comment]
Retest this please

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: nit: could use `filterNot`

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: I'd think the `group.is(Dead)` check should be the first thing we do.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: Not sure I get this. Why are we using the memberId as the purgatory key?

File: core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala]: nit: braces are not needed

File: core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala]: nit: maybe use `case None` since it is the only possibility.

Also, I think it might be worth having a log message somewhere indicating that the group cannot complete the rebalance because there is no leader joined.

File: core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala]: nit: this reads a bit odd. Wouldn't it make more sense to call this `currentLeader`?

**Review by hachikuji — COMMENTED** [review]
Thanks, left a few more comments.

File: core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala
Comment (abbccdda) [reply core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala]: Sounds good!

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (abbccdda) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: I think we could use a dummy key here.

Comment (abbccdda) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: In fact, I do realize the purpose here. If this is last delayed join expiration, we will not trigger `onJoinComplete` again only depending on heartbeat timeout. So we have to constantly check for group emptiness

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (abbccdda) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:792]: @hachikuji This test case covers the case where all static members are not joined but haven't session timeout yet. No new generation shall be bumped.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: We should add groupId to this message. Conventionally, the first word should be capitalized. Same for the log line above.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: Still not sure I get this. What is the point of using a different key?

File: core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala]: We actually return `None`, right? Possibly we could just have this method return a boolean since we are just checking `isDefined` on the result in `onCompleteJoin`.

File: core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala]: nit: conventionally, we would write this:
```scala
leaderId.flatMap { currentLeaderId =>
  ...
}
```

File: core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala]: We should be able to destructure `anyJoinedMember`. Something like this:
```scala
case Some((anyJoinedMemberId, anyJoinedMember)) =>
```

**Review by hachikuji — COMMENTED** [review]
Thanks for the updates. Just a few more comments.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (abbccdda) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: Yea you are right, we should use `GroupKey(group.groupId)`!

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (abbccdda) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: Sounds good!

File: core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala
Comment (abbccdda) [reply core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala]: Ack

File: core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala
Comment (abbccdda) [reply core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala]: Make sense

**Comment by abbccdda** [comment]
Retest this please

File: core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala]: nit: this use of `fold` is a tad odd. Perhaps we could use `leaderId.map{ ... }.getOrElse(false)`? Also, we don't need the second braces.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:794]: Should we assert that the number of members is 1?

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala]: I guess another case is if the static member rejoins instead of timing out. Do you think this is worth another test case?

**Review by hachikuji — COMMENTED** [review]
Thanks, just a couple more comments.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (abbccdda) [reply core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:794]: Sounds good

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (abbccdda) [reply core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala]: I think we have many `static member rejoining during PrepareRebalance` tests. This should be ok to skip.

Comment (hachikuji) [reply core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala]: Yeah, my thought is to exercise the "dummy" rebalance logic completely.

Comment (abbccdda) [reply core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala]: Sounds good, I got a new test case branching from the existing one @hachikuji

**Review by hachikuji — APPROVED** [review]
LGTM. Thanks for the patch.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:937]: Personally I'd prefer to have such logic in DelayedJoin instead, i.e. to re-write the current 

```
override def onComplete() = coordinator.onCompleteJoin(group)
```

code. Similar to the extended `InitialDelayedJoin#onComplete`.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (guozhangwang) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:483]: Not clear why we want to override to half of the value?

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (guozhangwang) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:525]: Ditto, why double the value here?

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (guozhangwang) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:800]: Thanks for adding these two test cases for improving the coverage!

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (guozhangwang) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:1177]: Nice refactoring.

**Review by guozhangwang — COMMENTED** [review]
LGTM! Just some minor questions.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (abbccdda) [reply core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:483]: We are just trying to make sure we timeout the follower instance because default session timeout = default rebalance timeout

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (abbccdda) [reply core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:525]: Doubling is aiming to reproduce the scenario where `rebalanceTImeout` < `sessionTimeout`. I could write comments for both scenarios.

**Pull request merged**
